src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -0,0 +1,370 @@
+            // The caller should catch all exceptions and swallow.
+
+            // First step: Get all the reflection objects we will ever need.
+            Assembly systemNetHttpAssembly = typeof(ServicePoint).Assembly;
+            s_connectionGroupListField = typeof(ServicePoint).GetField("m_ConnectionGroupList", BindingFlags.Instance | BindingFlags.NonPublic);
+            s_connectionGroupType = systemNetHttpAssembly?.GetType("System.Net.ConnectionGroup");
+            s_connectionListField = s_connectionGroupType?.GetField("m_ConnectionList", BindingFlags.Instance | BindingFlags.NonPublic);
+            s_connectionType = systemNetHttpAssembly?.GetType("System.Net.Connection");
+            s_writeListField = s_connectionType?.GetField("m_WriteList", BindingFlags.Instance | BindingFlags.NonPublic);
+
+            // Second step: Generate an accessor for HttpWebRequest._HttpResponse
+            FieldInfo field = typeof(HttpWebRequest).GetField("_HttpResponse", BindingFlags.NonPublic | BindingFlags.Instance);
+
+            string methodName = field?.ReflectedType.FullName + ".get_" + field.Name;
+            if (!string.IsNullOrEmpty(methodName))
+            {
+                DynamicMethod getterMethod = new DynamicMethod(methodName, typeof(HttpWebResponse), new Type[] { typeof(HttpWebRequest) }, true);
+                ILGenerator generator = getterMethod.GetILGenerator();
+                generator.Emit(OpCodes.Ldarg_0);
+                generator.Emit(OpCodes.Ldfld, field);
+                generator.Emit(OpCodes.Ret);
+                s_httpResponseAccessor = (Func<HttpWebRequest, HttpWebResponse>)getterMethod.CreateDelegate(typeof(Func<HttpWebRequest, HttpWebResponse>));
+            }
+
+            // Double checking to make sure we have all the pieces initialized
+            if (s_connectionGroupListField == null ||
+                s_connectionGroupType == null ||
+                s_connectionListField == null ||
+                s_connectionType == null ||
+                s_writeListField == null ||
+                s_httpResponseAccessor == null)
+            {
+                throw new InvalidOperationException("Unable to initialize all required reflection objects");
+            }
+        }
+
+        private static void PerformInjection()
+        {
+            FieldInfo servicePointTableField = typeof(ServicePointManager).GetField("s_ServicePointTable", BindingFlags.Static | BindingFlags.NonPublic);
+            if (servicePointTableField == null)
+            {
+                throw new InvalidOperationException("Unable to access the ServicePointTable field");
+            }
+
+            Hashtable originalTable = servicePointTableField.GetValue(null) as Hashtable;
+            ServicePointHashtable newTable = new ServicePointHashtable();
+
+            // Copy the existing entries over to the new table, and replace the field with our new table
+            if (originalTable != null)
```

Comment:
We are racing with anything that is updating s_ServicePointTable while we are prepping our update here.   What kinds of updates happen to this table?   Do things leave get removed from this table?   Is it possible to do a 'double check' (after installing insuring that there are no 'stragglers' that are in the old table but not the new one).   This is the main place where we can 'break' things.   

We should at least document that we understand this race and view it as the best of a bad situation.